avalanche/distributed/distributed_batch.py

```diff
@@ -0,0 +1,129 @@
+from abc import abstractmethod, ABC
+from typing import TypeVar, List, Optional, Callable, Any, Tuple
+
+import torch
+from torch import Tensor
+
+from avalanche.distributed import DistributedHelper
+from avalanche.distributed.distributed_value import SwitchableDistributedValue
+
+LocalT = TypeVar('LocalT')
+DistributedT = TypeVar('DistributedT')
+
+
+class DistributedObject(SwitchableDistributedValue[LocalT, DistributedT], ABC):
+    """
+    An intermediate abstract class in charge of synchronizing objects.
+
+    The merge procedure must be implemented in child classes.
+    """
+    def _synchronize_distributed_value(self) -> DistributedT:
+        objects = self._synchronize_objects()
+        return self._merge_objects(objects)
+
+    def _synchronize_objects(self) -> List[LocalT]:
+        return DistributedHelper.gather_all_objects(
+            self._local_value
+        )
+
+    @abstractmethod
+    def _merge_objects(self, objects: List[LocalT]) -> DistributedT:
+        pass
+
+
+class DistributedBatch(DistributedObject[LocalT, LocalT], ABC):
+    """
+    An intermediate abstract class in charge of synchronizing data batches.
+
+    This class can handle batches as either tuples of elements (as usual) or
+    even single values.
```

Comment:
I don't understand this. Why do we need a single class to deal with both tuples and single values? It seems easier to have different classes each with their own merge and sync methods.

Reply:
A class that can handle both situations may be useful for unsupervised strategies that may have a batch made of the X value only. In that case, the batch is not a tuple and things need to be managed in a slightly different way. Switching classes depending on that may be more complicated than doing this...